[tuya] Implemented missing wifi module-related commands

szupi-ipuzs opened the pull request:
# What does this implement/fix?

I have a Vactidy T6 which works with tuya (standard serial protocol) and i can see it spamming the wifi module with command 0x2D ("Get MAC address") and 0x34/0x07 (the Extended "Get module information" command). Their implementation is currently missing so I'm adding it.
It is currently unclear to me what the mcu needs these information for. Even without sending the response, the robot was responding to other commands. Still this might differ depending on the firmware, so it's better to send and answer anyway.

My robot's original module (WBR3) has been replaced with an ESP-12F and thus its MAC differs from the original. I send the new mac in answer to the 0x2D command and the robot seems to accept it (although it never stops requesting it while in ON state). Still for cases like that we might consider adding allowing the user to provide a custom fake MAC via yaml....

The 0x34/0x07 is sent by my robot only to request the AP ssid (the one used by tuya AP pairing mode). As above, it actually floods us with this command while the robot is ON. It is unclear if we should provide it only when in AP mode or always? Anyway, my implementation always sends the constant string "smartlife".

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

szupi-ipuzs committed:
Implemented missing commands

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

szupi-ipuzs committed:
Use raw string literals

bdraco committed:
Merge branch 'dev' into tuya_module_info_commands

bdraco commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -335,6 +342,24 @@ void Tuya::handle_command_(uint8_t command, uint8_t version, const uint8_t *buff
         }
+        case TuyaExtendedServicesCommandType::GET_MODULE_INFORMATION: {
+          std::vector<uint8_t> response_payload;
+          std::string module_info_str;
```
It looks like `send_raw_command_` is synchronous. We should fix `send_raw_command_` to have a uint8_t overload before adding more code that heap allocates

bdraco commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -316,6 +316,13 @@ void Tuya::handle_command_(uint8_t command, uint8_t version, const uint8_t *buff
     }
+    case TuyaCommandType::GET_MAC_ADDRESS: {
+      std::vector<uint8_t> mac(6u);
+      get_mac_address_raw(mac.data());
```
We should use the non-heap allocating version of this

bdraco committed:
[tuya] Eliminate heap allocations in module info and MAC address commands

Add send_raw_command_ overload taking raw uint8_t*/size_t to avoid
constructing std::vector for synchronous UART writes. Rewrite
process_get_module_information_ to build JSON directly into a stack
buffer instead of returning std::string.

bdraco committed:
[tuya] Use ternary for module info success/failure status

bdraco committed:
[tuya] Use MAC_ADDRESS_SIZE and MAC_ADDRESS_PRETTY_BUFFER_SIZE constexprs

bdraco commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -760,5 +799,68 @@ void Tuya::register_listener(uint8_t datapoint_id, const std::function<void(Tuya
+  size_t pos = 0;
+  auto append = [&](const char *str) { pos = buf_append_bytes(output, output_size, pos, str); };
+
+  append("{");
```
all the ram increase is driven by the strings stored in rodata on esp8266.  could use `buf_append_printf` as it auto stores them in flash but likely we should add another helper for that since there are just copying.

Too tired to work that out tonight though
